File: SourceAFIS/Simple/Person.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Text;
using System.Xml.Serialization;
using System.Linq;
using SourceAFIS.Dummy;

namespace SourceAFIS.Simple
{
    /// <summary>
    /// Collection of <see cref="Fingerprint"/>s belonging to one person.
    /// </summary>
    /// <remarks>
    /// <para>
    /// This class is primarily a way to group multiple <see cref="Fingerprint"/>s belonging to one person.
    /// This is very convenient feature when there are multiple fingerprints per person, because
    /// it is possible to match two <see cref="Person"/>s directly instead of iterating over their <see cref="Fingerprint"/>s.
    /// </para>
    /// <para>
    /// <see cref="Id"/> field is provided as simple means to bind <see cref="Person"/> objects to application-specific
    /// information. If you need more flexibility, inherit from <see cref="Person"/> class and add
    /// application-specific fields as necessary.
    /// </para>
    /// <para>
    /// This class is designed to be easy to serialize in order to be stored in binary format (BLOB)
    /// in application database, binary or XML files, or sent over network. You can either serialize
    /// the whole <see cref="Person"/> or serialize individual <see cref="Fingerprint"/>s.
    /// </para>
    /// </remarks>
    /// <seealso cref="Fingerprint"/>
    [Serializable]
    public class Person : ICloneable
    {
        /// <summary>
        /// Application-assigned ID for the <see cref="Person"/>.
        /// </summary>
        /// <remarks>
        /// <para>
        /// SourceAFIS doesn't use this property. It is provided for applications as an easy means
        /// to link <see cref="Person"/> objects back to application-specific data. Applications can store any
        /// integer ID in this field, for example database table key or an array index.
        /// </para>
        /// <para>
        /// Applications that need to attach more detailed information to the person should
        /// inherit from <see cref="Person"/> class and add fields as necessary.
        /// </para>
        /// </remarks>
        [XmlAttribute]
        public int Id { get; set; }

        List<Fingerprint> FingerprintList = new List<Fingerprint>();

        /// <summary>
        /// List of <see cref="Fingerprint"/>s belonging to the <see cref="Person"/>.
        /// </summary>
        /// <remarks>
        /// This collection is initially empty. Add <see cref="Fingerprint"/> objects
        /// here. You can also assign the whole collection.
        /// </remarks>
        public List<Fingerprint> Fingerprints
        {
            get { return FingerprintList; }
            set
            {
                if (value == null)
                    throw new ArgumentNullException();
                FingerprintList = value;
            }
        }

        /// <summary>
        /// Creates empty <see cref="Person"/> object.
        /// </summary>
        public Person()
        {
        }

        /// <summary>
        /// Creates new <see cref="Person"/> object and initializes it with
        /// a list of <see cref="Fingerprint"/>s.
        /// </summary>
        /// <param name="fingerprints"><see cref="Fingerprint"/> objects to add to the new <see cref="Person"/>.</param>
        public Person(params Fingerprint[] fingerprints)
        {
            Fingerprints = fingerprints.ToList();
        }

        /// <summary>
        /// Create deep copy of the <see cref="Person"/>.
        /// </summary>
        /// <returns>Deep copy of the <see cref="Person"/>.</returns>
        /// <remarks>
        /// This method also clones all <see cref="Fingerprint"/> objects contained
        /// in this <see cref="Person"/>.
        /// </remarks>
        public Person Clone()
        {
            Person clone = new Person();
            clone.Id = Id;
            foreach (Fingerprint fp in Fingerprints)
                clone.Fingerprints.Add(fp.Clone());
            return clone;
        }

        object ICloneable.Clone() { return Clone(); }

        internal void CheckForNulls()
        {
            foreach (Fingerprint fp in Fingerprints)
                if (fp == null)
                    throw new ApplicationException("Person contains null Fingerprint references.");
        }
    }
}

File: java/sourceafis/src/sourceafis/simple/Person.java
package sourceafis.simple;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

 /**
  * Collection of {@link Fingerprint}s belonging to one person.
  * 
  * This class is primarily a way to group multiple {@link Fingerprint}s belonging to one person.
  * This is very convenient feature when there are multiple fingerprints per person, because
  * it is possible to match two {@link Person}s directly instead of iterating over their {@link Fingerprint}s.
  * <p>
  * {@link #setId Id} property is provided as a simple means to bind {@link Person} objects to application-specific
  * information. If you need more flexibility, inherit from {@link Person} class and add
  * application-specific fields as necessary.
  * <p>
  * This class is designed to be easy to serialize in order to be stored in binary format (BLOB)
  * in application database, binary or XML files, or sent over network. You can either serialize
  * the whole {@link Person} or serialize individual {@link Fingerprint}s.
  * 
  * @see Fingerprint 
  * @serial exclude
  */
 
@SuppressWarnings("serial")
public class Person implements Cloneable,Serializable
{
    private int Id;
    private List<Fingerprint> fingerprints = new ArrayList<Fingerprint>();

    /**
     * Creates an empty {@code Person} object.
     */
    public Person() { }

    /**
     * Creates new {@code Person} object and initializes it with a list of
     * {@link Fingerprint}s.
     * 
     * @param fingerprints
     *            {@link Fingerprint} objects to add to the new {@code Person}
     */
    public Person(Fingerprint... fingerprints) {
    	for(Fingerprint fp:fingerprints){
    		this.fingerprints.add(fp);
    	}
        //this.Fingerprints =  fingerprints.ToList();
    }

    /**
     * Gets application-defined ID for the Person.
     * 
     * See {@link #setId setId} for explanation. This method just returns
     * previously set ID.
     * 
     * @return ID that was previously set via {@link #setId setId}
     * @see #setId setId
     */
    public int getId() {
		return Id;
	}
    
    /**
     * Sets application-defined ID for the {@code Person}.
     * 
     * SourceAFIS doesn't use this ID. It is provided for applications as an
     * easy means to link {@code Person} objects back to application-specific
     * data. Applications can store any integer ID in this field, for example
     * database table key or an array index.
     * <p>
     * Applications that need to attach more detailed information to the person
     * should inherit from {@code Person} class and add fields as necessary.
     * 
     * @param id
     *            arbitrary application-defined ID
     * @see #getId getId
     */
    public void setId(int id) {
        Id = id;
    }
    
    /**
     * Gets list of {@link Fingerprint}s belonging to the {@link Person}.
     * 
     * This collection is initially empty. Add {@link Fingerprint} objects
     * to the returned collection.
     * 
     * @see #setFingerprints setFingerprints
     * @see Fingerprint
     */
    public List<Fingerprint> getFingerprints() {
        return fingerprints;
    }

    /**
     * Sets list of {@link Fingerprint}s belonging to the {@link Person}.
     * 
     * You can assign the whole collection using this method. Individual
     * {@link Fingerprint}s can be added to the collection returned from
     * {@link #getFingerprints getFingerprints}.
     * 
     * @param fingerprints
     *            new list of {@link Fingerprint}s for this {@code Person}
     * @see #getFingerprints getFingerprints
     * @see Fingerprint
     */
    public void setFingerprints(List<Fingerprint> fingerprints) {
        this.fingerprints = fingerprints;
    }

    /**
     * Creates deep copy of the Person.
     * 
     * This method clones all {@link Fingerprint} objects contained in this
     * {@code Person}.
     * 
     * @return deep copy of the {@code Person}
     */
    public Person clone() {
	    Person copy = new Person();
	    copy.Id = Id;
	    for (Fingerprint fingerprint : fingerprints)
	        copy.fingerprints.add(fingerprint.clone());
		return copy;
    }

    void CheckForNulls()
    {
        for (Fingerprint fp : fingerprints)
            if (fp == null)
                throw new RuntimeException("Person contains null Fingerprint references.");
    }
}
